CT-1245: upload in background

BobHemphill opened the pull request:
Here's my start to getting the native Ios uploader to work in the background...I need to integrate with @adamtysonsmith and @aacotroneo to test this

BobHemphill committed:
CT-1245: upload in background

CT-1245: cleanup interface

CT_1245: background

CT-1245: clean up uploader with filemanager

CT-1245: amazon header

CT-1245: putting error back in

CT-1245: body

CT-1245: removing download patch in Sebs PR

CT-1245: integrating adams background flag

CT-1245: spike for background processing

CT-1245: integration bool value from reactnatve call

CT-1245: logs

CT-1245: more logs

CT-1245: background only

CT-1245: suspension bg processing

CT-1245: NSLog commits

CT-1245: options in react native sent to bridge

CT-1245: getting body from lifecycle event

CT-1245: comments to ensure body

CT-1245: response dict fun

k7k0 commented on Downloader.m:
```diff
@@ -88,6 +88,10 @@ - (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTas
 {
+    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)downloadTask.response;
+    if (!_statusCode) {
+        _statusCode = [NSNumber numberWithLong:httpResponse.statusCode];
```
badass!! seems to work in suspended: `didFinishDownloadingToURL. Status: 200`

k7k0 replied:
I've added the download status code fetching in my own PR, https://github.com/CommercialTribe/react-native-fs/pull/4
 maybe you could remove it from this PR to avoid conflicts

k7k0 commented on Uploader.m:
```diff
@@ -24,85 +24,25 @@ - (void)uploadFiles:(RNFSUploadParams*)params
   // set headers
-  NSString *formBoundaryString = [self generateBoundaryString];
-  NSString *contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@", formBoundaryString];
+  NSString *contentType = @"video/mp4";
```
we will probably upload thumbnails too (we're sending a `Content-type` header in the @aacotroneo merge)

k7k0 commented on Uploader.m:
```diff
@@ -24,85 +24,25 @@ - (void)uploadFiles:(RNFSUploadParams*)params
+  if (![fileManager fileExistsAtPath:filepath]){
+    NSLog(@"Failed to open target file at path: %@", filepath);
+    NSError* error = [NSError errorWithDomain:@"Uploader" code:NSURLErrorFileDoesNotExist userInfo:@{NSLocalizedDescriptionKey: [NSString stringWithFormat: @"Failed to open target file at path: %@", filepath]}];
+    return _params.errorCallback(error);
```
great! I dont know why they commented the error callback in the first place... doesnt make sense to me

BobHemphill replied:
it's commented out because this code doesn't work if the file url/path is rooted at file:///

k7k0 reviewed (approved):
Really cool! Maybe we could merge so I could use the download stuff you've added

BobHemphill commented:
We can probably get this merged on Monday. I've been experimenting with the
ale bg branch and successfully got an upload using the background upload
task and delegates. I wasn't able to test background uploading just yet but
it's getting close.

Oh yeah and the content type I just added before I rebased with alejandros
branch. I need to clean up with Adams changes and integrate with alejandros
branch and then test the uploads in the background. The last piece is
tieing in the completion handler stuff

On Friday, September 23, 2016, Sebastian Marconi notifications@github.com
wrote:

> _@k7k0_ approved this pull request.
> 
> Really cool! Maybe we could merge so I could use the download stuff you've
> 
> ## added
> 
> In Uploader.m
> https://github.com/CommercialTribe/react-native-fs/pull/3#pullrequestreview-1413445
> :
> 
> > @@ -24,85 +24,25 @@ - (void)uploadFiles:(RNFSUploadParams*)params
> >    [req setHTTPMethod:method];
> 
>    // set headers
> -  NSString *formBoundaryString = [self generateBoundaryString];
> -  NSString *contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@", formBoundaryString];
> -  NSString *contentType = @"video/mp4";
> 
> we will probably upload thumbnails too (we're sending a Content-type
> 
> ## header in the @aacotroneo https://github.com/aacotroneo merge)
> 
> In Uploader.m
> https://github.com/CommercialTribe/react-native-fs/pull/3#pullrequestreview-1413445
> :
> 
> > -    } else {
> > -      [reqBody appendData:[[NSString stringWithFormat:@"Content-Type: %@\r\n", [self mimeTypeForPath:filename]] dataUsingEncoding:NSUTF8StringEncoding]];
> > 
> > ## \-    }
> > -    [reqBody appendData:[[NSString stringWithFormat:@"Content-Length: %ld\r\n\r\n", (long)[fileData length]] dataUsingEncoding:NSUTF8StringEncoding]];
> > -    [reqBody appendData:fileData];
> > -    [reqBody appendData:[@"\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
> > -  // add file
> > -  NSDictionary *file = [_params.files objectAtIndex:0];
> > -  NSString *filepath = file[@"filepath"];
> >   +
> > -  NSFileManager *fileManager = [NSFileManager defaultManager];
> > -  if (![fileManager fileExistsAtPath:filepath]){
> > -    NSLog(@"Failed to open target file at path: %@", filepath);
> > -    NSError\* error = [NSError errorWithDomain:@"Uploader" code:NSURLErrorFileDoesNotExist userInfo:@{NSLocalizedDescriptionKey: [NSString stringWithFormat: @"Failed to open target file at path: %@", filepath]}];
> > -    return _params.errorCallback(error);
> 
> great! I dont know why they commented the error callback in the first
> place... doesnt make sense to me
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/CommercialTribe/react-native-fs/pull/3#pullrequestreview-1413445,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AAstIZ92babkn0DrJLa0XGlKjG-wMXJuks5qtE-ggaJpZM4KFam_
> .

k7k0 commented on Uploader.m:
```diff
@@ -50,104 +85,119 @@ - (void)uploadFiles:(RNFSUploadParams*)params
     return _params.errorCallback(error);
   }
+    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)task.response;
+    NSString *body = @"rock your body";
```
is this a TODO ?

BobHemphill replied:
Whoops, we actually don't even need that body...aside from it can't be nil because it breaks the completeHandler code in the bridge when it's nil.  But our promise chain doesn't even use the upload response.

This also, isnt the end of this story.  Because there's a lot that needs to change to background tasks.

BobHemphill replied:
But I can turn that into a todo...there's a different lifecycle event we need to handle to recieve body data from the response...

BobHemphill replied:
added response data handling
